README.md

```diff
@@ -0,0 +1,217 @@
+
+## Objectives
+  1. Explain why, exactly, it's a bad idea to store passwords in plaintext.
+  2. Write code to store and verify hashed, salted passwords.
+  3. Use Rails' `has_secure_password` to store and authenticate users login credentials securely.
+
+## The problem with passwords
+
+Let's imagine a `SessionsController#create` method that does very simple authentication. It goes like this:
+
+```ruby
+   def create
+     @user = User.find(username: params[:username])
+     return head(:forbidden) unless params[:password] == @user.password
+     session[:user_id] = @user.id
+   end
+```
+
+We load the user row, check to see if the provided password is equal to the password stored in the database, and if it is, we set `user_id` in the `session`.
+
+This is tremendously insecure.
+
+The reason this is tremendously insecure is that you have to store all your users' passwords in the database, unencrypted.
+
+Never do this.
+
+Even if you don't care about the security of your site, people have a strong tendency to reuse passwords. That means that the inevitable security breach of your site will leak passwords which some users also use for Gmail. Your `users` table probably has an `email` column. This means that if I'm a hacker, getting access to your database has given me the Internet equivalent of the house keys and home address for some (probably surprisingly large) percentage of your users.
+
+## Hashing passwords
+
+So how do we store passwords if we can't store passwords?
+
+We store their hashes. A *hash* is a number computed by feeding a string to a *hash function*. Hash functions have the property that they will always produce the same number given the same input. You could write one yourself. Here's one that I just made:
+
+```ruby
+   # dumb_hash(input: string) -> number
+   def dumb_hash(input)
+     input.bytes.reduce(:+)
+   end
+```
+
+My `dumb_hash` function just finds the sum of the bytes that comprise the string. It is a hash function, since it satisfies the criteria that the same string always produces the same result.
+
+We could imagine using this function to avoid storing passwords in the database. Our `User` model and `SessionsController` might look like this:
+
+```ruby
+   # app/models/user.rb
+   class User < ActiveRecord::Base
+     def password=(new_password)
+       self.password_digest = dumb_hash(new_password)
+     end
+
+     def authenticate(password)
+       return nil unless dumb_hash(password) == password_digest
+       self
+     end
+
+     private
+
+     def dumb_hash(input)
+       input.bytes.reduce(:+)
+     end
+   end
+
+   # app/controllers/sessions_controller.rb
+   class SessionsController < ApplicationController
+     def create
+       user = User.find_by(username: params[:username])
+       authenticated = user.try(:authenticate, params[:pasword])
```

Comment:
is this the first time we're using `.try`? if so, let's explain is briefly with the doc: http://api.rubyonrails.org/classes/Object.html#method-i-try